<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Bioconductor | 연구데이터 분석과정 (R프로그래밍)</title>
  <meta name="description" content="Lecture" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Bioconductor | 연구데이터 분석과정 (R프로그래밍)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Lecture" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Bioconductor | 연구데이터 분석과정 (R프로그래밍)" />
  
  <meta name="twitter:description" content="Lecture" />
  

<meta name="author" content="한국생명공학연구원 김하성" />


<meta name="date" content="2020-10-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="basic-statistics.html"/>
<link rel="next" href="data-science-and-tidyverse.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">[2020 KRIBB Lecture] R programming</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction 강의 개요</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#references-books-참고-교제"><i class="fa fa-check"></i><b>1.1</b> References books 참고 교제</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#references-참고-자료"><i class="fa fa-check"></i><b>1.2</b> References 참고 자료</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#schedule-강의-계획"><i class="fa fa-check"></i><b>1.3</b> Schedule 강의 계획</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html"><i class="fa fa-check"></i><b>2</b> R/Rstudio basics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#what-is-r-rstudio"><i class="fa fa-check"></i><b>2.1</b> What is R / Rstudio</a></li>
<li class="chapter" data-level="2.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#r-rstudio-installation"><i class="fa fa-check"></i><b>2.2</b> R / Rstudio installation</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#r-설치"><i class="fa fa-check"></i><b>2.2.1</b> R 설치</a></li>
<li class="chapter" data-level="2.2.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#rstudio-설치"><i class="fa fa-check"></i><b>2.2.2</b> Rstudio 설치</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#rstudio-interface"><i class="fa fa-check"></i><b>2.3</b> Rstudio interface</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#keyboard-shortcuts"><i class="fa fa-check"></i><b>2.3.1</b> Keyboard shortcuts</a></li>
<li class="chapter" data-level="2.3.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#set-a-working-directory"><i class="fa fa-check"></i><b>2.3.2</b> Set a working directory</a></li>
<li class="chapter" data-level="2.3.3" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#set-a-project"><i class="fa fa-check"></i><b>2.3.3</b> Set a project</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#r-coding-practice"><i class="fa fa-check"></i><b>2.4</b> R coding practice</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#console-calculator"><i class="fa fa-check"></i><b>2.4.1</b> Console calculator</a></li>
<li class="chapter" data-level="2.4.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#variables-and-values"><i class="fa fa-check"></i><b>2.4.2</b> Variables and values</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#supports"><i class="fa fa-check"></i><b>2.5</b> Supports</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#help"><i class="fa fa-check"></i><b>2.5.1</b> Help</a></li>
<li class="chapter" data-level="2.5.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#cheatsheet"><i class="fa fa-check"></i><b>2.5.2</b> Cheatsheet</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#r-packages-and-dataset"><i class="fa fa-check"></i><b>2.6</b> R packages and Dataset</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#r-packages"><i class="fa fa-check"></i><b>2.6.1</b> R packages</a></li>
<li class="chapter" data-level="2.6.2" data-path="rrstudio-basics.html"><a href="rrstudio-basics.html#data-sets"><i class="fa fa-check"></i><b>2.6.2</b> Data sets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-statistics.html"><a href="basic-statistics.html"><i class="fa fa-check"></i><b>3</b> Basic Statistics</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-statistics.html"><a href="basic-statistics.html#univariat-data"><i class="fa fa-check"></i><b>3.1</b> Univariat data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-statistics.html"><a href="basic-statistics.html#vectors-for-univariate-data"><i class="fa fa-check"></i><b>3.1.1</b> Vectors for univariate data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-statistics.html"><a href="basic-statistics.html#storage-data-type"><i class="fa fa-check"></i><b>3.1.2</b> Storage data type</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-statistics.html"><a href="basic-statistics.html#use-of-functions"><i class="fa fa-check"></i><b>3.1.3</b> Use of functions</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-statistics.html"><a href="basic-statistics.html#numeric-summaries"><i class="fa fa-check"></i><b>3.1.4</b> Numeric summaries</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-statistics.html"><a href="basic-statistics.html#center-for-a-univariat-variable"><i class="fa fa-check"></i><b>3.1.5</b> Center for a univariat variable</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-statistics.html"><a href="basic-statistics.html#spread-for-a-univariat-variable"><i class="fa fa-check"></i><b>3.1.6</b> Spread for a univariat variable</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-statistics.html"><a href="basic-statistics.html#shape-for-a-univariat-variable"><i class="fa fa-check"></i><b>3.1.7</b> Shape for a univariat variable</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-statistics.html"><a href="basic-statistics.html#viewing-the-shape"><i class="fa fa-check"></i><b>3.1.8</b> Viewing the shape</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="basic-statistics.html"><a href="basic-statistics.html#bivariate-data"><i class="fa fa-check"></i><b>3.2</b> Bivariate data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="basic-statistics.html"><a href="basic-statistics.html#independence-samples"><i class="fa fa-check"></i><b>3.2.1</b> Independence samples</a></li>
<li class="chapter" data-level="3.2.2" data-path="basic-statistics.html"><a href="basic-statistics.html#data-comparison-with-plots"><i class="fa fa-check"></i><b>3.2.2</b> Data comparison with plots</a></li>
<li class="chapter" data-level="3.2.3" data-path="basic-statistics.html"><a href="basic-statistics.html#data-manipulation"><i class="fa fa-check"></i><b>3.2.3</b> Data manipulation</a></li>
<li class="chapter" data-level="3.2.4" data-path="basic-statistics.html"><a href="basic-statistics.html#paired-data"><i class="fa fa-check"></i><b>3.2.4</b> Paired data</a></li>
<li class="chapter" data-level="3.2.5" data-path="basic-statistics.html"><a href="basic-statistics.html#bivariate-categorical-data"><i class="fa fa-check"></i><b>3.2.5</b> Bivariate categorical data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="bioconductor.html"><a href="bioconductor.html"><i class="fa fa-check"></i><b>4</b> Bioconductor</a>
<ul>
<li class="chapter" data-level="4.1" data-path="bioconductor.html"><a href="bioconductor.html#working-with-objects"><i class="fa fa-check"></i><b>4.1</b> Working with objects</a></li>
<li class="chapter" data-level="4.2" data-path="bioconductor.html"><a href="bioconductor.html#discovering-installing-and-learning-how-to-use-bioconductor-packages"><i class="fa fa-check"></i><b>4.2</b> Discovering, installing, and learning how to use Bioconductor packages</a></li>
<li class="chapter" data-level="4.3" data-path="bioconductor.html"><a href="bioconductor.html#data-structure-and-management-for-genome-scale-experiments"><i class="fa fa-check"></i><b>4.3</b> Data structure and management for genome-scale experiments</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="bioconductor.html"><a href="bioconductor.html#expressionset-class"><i class="fa fa-check"></i><b>4.3.1</b> ExpressionSet Class</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="bioconductor.html"><a href="bioconductor.html#biostrings-package-for-sequence-analysis"><i class="fa fa-check"></i><b>4.4</b> Biostrings package for sequence analysis</a></li>
<li class="chapter" data-level="4.5" data-path="bioconductor.html"><a href="bioconductor.html#download-sequence-information-from-ncbi"><i class="fa fa-check"></i><b>4.5</b> Download sequence information from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html"><i class="fa fa-check"></i><b>5</b> Data science and tidyverse</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#tidy-data-structure"><i class="fa fa-check"></i><b>5.1</b> Tidy data structure</a></li>
<li class="chapter" data-level="5.2" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#analysis-goal-all-dataset"><i class="fa fa-check"></i><b>5.2</b> Analysis Goal (ALL dataset)</a></li>
<li class="chapter" data-level="5.3" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#apply-functions"><i class="fa fa-check"></i><b>5.3</b> Apply functions</a></li>
<li class="chapter" data-level="5.4" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#data-manipulation-with-dplyr"><i class="fa fa-check"></i><b>5.4</b> Data manipulation with dplyr</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#dplyr---pipe-operator"><i class="fa fa-check"></i><b>5.4.1</b> dplyr - pipe operator</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#dplyr---important-functions"><i class="fa fa-check"></i><b>5.4.2</b> dplyr - Important functions</a></li>
<li class="chapter" data-level="5.4.3" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#code-comparison"><i class="fa fa-check"></i><b>5.4.3</b> code comparison</a></li>
<li class="chapter" data-level="5.4.4" data-path="data-science-and-tidyverse.html"><a href="data-science-and-tidyverse.html#dplyr-example-iris"><i class="fa fa-check"></i><b>5.4.4</b> dplyr example iris</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="analysis-example.html"><a href="analysis-example.html"><i class="fa fa-check"></i><b>6</b> Analysis example</a>
<ul>
<li class="chapter" data-level="6.1" data-path="analysis-example.html"><a href="analysis-example.html#all-dataset"><i class="fa fa-check"></i><b>6.1</b> ALL dataset</a></li>
<li class="chapter" data-level="6.2" data-path="analysis-example.html"><a href="analysis-example.html#heatmap-with-clustering"><i class="fa fa-check"></i><b>6.2</b> Heatmap with clustering</a></li>
<li class="chapter" data-level="6.3" data-path="analysis-example.html"><a href="analysis-example.html#smooth-line"><i class="fa fa-check"></i><b>6.3</b> Smooth line</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">연구데이터 분석과정 (R프로그래밍)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bioconductor" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Bioconductor</h1>
<p>Bioconductor는 바이오인포메틱스를 위한 R기반의 데이터, 메소드, 그리고 패키지들의 모음입니다. 2002년 microarray 데이터 분석을 위한 플랫폼으로 시작되었으며 현재 1,300개 이상의 패키지로 구성되어 있습니다. R은 분산형 오픈소스이나 Bioconductor는 Full-time developer들에 의해서 유지되고 있습니다. <code>CRAN</code>에 배포되지 않고 <code>CRAN</code>에 비해 더 많은 필수 자료들 (vignettes 등)이 필요하며 높은 수준으로 quality control이 되고 있습니다. 현재 RNA-seq, ChIP seq, copy number analysis, microarray methylation, classic expression analysis, flow cytometry 등 다양한 분야의 데이터 분석에 사용되고 있습니다.</p>
<p><img src="images/04/bioconductor.JPG" /></p>
<p><strong>특징</strong></p>
<p>Bioconductor 코어 개발 그룹은 사용자들이 지놈스케일 데이터를 더 편리하게 다루룰 수 있도록 데이터의 구조를 개발하는데 많은 시간을 들입니다.</p>
<ul>
<li>지놈스케일의 서열이나 발현등 대용량 데이터 관리<br />
</li>
<li>통계적 분석을 용이하게 수행하기 위한 전처리<br />
</li>
<li>분자수준의 현상과 생장이나 질병 등 표현형수준의 관계를 규명하기 위한 정량 데이터 통합<br />
</li>
<li>재사용 가능한 데이터를 위한 관리</li>
</ul>
<div id="working-with-objects" class="section level2">
<h2><span class="header-section-number">4.1</span> Working with objects</h2>
<p>객체지향프로그래밍 (OOP)은 복잡한 문제를 프로그래밍할 때 발생되는 코드의 복잡성을 해결할 수 있는 하나의 방안으로 1990년대부터 많이 사용되었습니다. Genome 스케일의 experiment나 annotation은 대표적인 복잡한 데이터 중 하나 입니다. <code>bioconcuctor</code>는 OOP의 개념을 도입하여 이러한 생물학 데이터를 구조화하고 효율적으로 데이터를 관리하고 있습니다.</p>
<p><img src="images/04/oop.png" /></p>
<p>Class와 Object, Instance 개념을 먼저 이해할 필요가 있습니다. 엄밀히 따지면 Object는 우리가 구현할 대상, Class는 설계도, Instance는 실제 구현한 대상 이라고 이해하시면 좋지만 가끔 Instance와 Object는 같은 의미로 사용되기도 합니다. 예를 들어 우리가 연구하는 <code>사람</code>, <code>원숭이</code>, <code>제브라피쉬</code> 라는 object들의 정보를 저장한다고 생각해 봅니다. 이 경우 각 객체들이 공통적으로 <code>name</code>, <code>age</code>, <code>genome</code> 등의 정보를 가지고 있고 그 기능적 특징 또한 유사하므로 이러한 정보가 저장되는 공간을 갖는 <code>생명체</code>라는 개념적 설계도를 (Class) 먼저 만듭니다. 그리고 해당 클래스의 <em>H.sapiens</em>, <em>D. rerio</em>, <em>B. Ape</em>, <em>C. goeldii</em> 를 만들면 각 객체의 구성 요소를 일일히 구현할 필요 없이 필요한 정보를 미리 정해진 구조에 맞게 저장 해주면 됩니다.</p>
<blockquote>
<p>Bioconductor에서 OOP 개념은 다음과 같습니다.<br />
class - 복잡한 데이터 구조의 틀<br />
object - 특정 클래스가 특정 구현된 실례 (instance)<br />
method - 특정 클래스의 구현된 기능 (get/set)</p>
</blockquote>
<p>그런데 이러한 class가 무수히 존재하며 각 name, age 등의 정보에 접근할 수 있는 method들을 제공하고 있습니다. class마다 사용 가능한 method가 어떠한 정보가 있는지 알기 위해서 <code>methods()</code>라는 함수를 사용합니다. 예를 들어 객체 <code>Homo.sapience</code>를 살펴보면 다음과 같습니다.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="bioconductor.html#cb66-1"></a><span class="kw">library</span>(Homo.sapiens)</span>
<span id="cb66-2"><a href="bioconductor.html#cb66-2"></a><span class="kw">class</span>(Homo.sapiens)</span>
<span id="cb66-3"><a href="bioconductor.html#cb66-3"></a>?OrganismDb</span></code></pre></div>
<blockquote>
<p>The OrganismDb class is a container for storing knowledge about existing Annotation packages and the relationships between these resources.</p>
</blockquote>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="bioconductor.html#cb67-1"></a><span class="kw">methods</span>(<span class="dt">class=</span><span class="kw">class</span>(Homo.sapiens))</span>
<span id="cb67-2"><a href="bioconductor.html#cb67-2"></a><span class="kw">genes</span>(Homo.sapiens)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb67-3"><a href="bioconductor.html#cb67-3"></a><span class="kw">exons</span>(Homo.sapiens)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb67-4"><a href="bioconductor.html#cb67-4"></a>homo_seq &lt;-<span class="st"> </span><span class="kw">seqinfo</span>(Homo.sapiens)</span>
<span id="cb67-5"><a href="bioconductor.html#cb67-5"></a><span class="kw">class</span>(homo_seq)</span>
<span id="cb67-6"><a href="bioconductor.html#cb67-6"></a>?Seqinfo</span></code></pre></div>
<blockquote>
<p>A Seqinfo object is a table-like object that contains basic information about a set of genomic sequences. …</p>
</blockquote>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="bioconductor.html#cb68-1"></a><span class="kw">length</span>(homo_seq)</span>
<span id="cb68-2"><a href="bioconductor.html#cb68-2"></a><span class="kw">seqnames</span>(homo_seq)</span></code></pre></div>
<p>bioconductor에는 대용량 정보가 object 형태로 구조화되어 저장되어 있으며 <code>library()</code>함수로 읽어올 수 있고 다양한 함수로 해당 object의 정보를 읽어올 수 있습니다.</p>
</div>
<div id="discovering-installing-and-learning-how-to-use-bioconductor-packages" class="section level2">
<h2><span class="header-section-number">4.2</span> Discovering, installing, and learning how to use Bioconductor packages</h2>
<p><a href="https://www.bioconductor.org" class="uri">https://www.bioconductor.org</a></p>
<p><strong>Discovery</strong></p>
<p>Use &gt;&gt; Software, Annotation, Experiment</p>
<ul>
<li>Software: 분석을 위한 툴 모음</li>
<li>Annotation: 유전자 symbol/ID mapping, gene ontology 기반 유전자 분류, 유전체상에서 exon, transcript, gene 등의 위치, 단백질 기능 등</li>
<li>Experiment data: 학습 가능할 정도의 Highly curated datasets (실험 데이터)</li>
<li>Workflow: 특정 데이터 분석을 위한 프로세스 모음</li>
</ul>
<p><strong>Installation</strong></p>
<p><code>BiocManager</code>를 먼저 설치하고 해당 패키지를 설치하시기 바랍니다. <code>BiocManager</code>에는 <code>available()</code>이라는 함수로 (특정 문자가 포함된) 사용 가능한 패키지를 검색할 수 도 있습니다.</p>
<p>Use &gt;&gt; Software &gt;&gt; IRanges</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="bioconductor.html#cb69-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb69-2"><a href="bioconductor.html#cb69-2"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb69-3"><a href="bioconductor.html#cb69-3"></a></span>
<span id="cb69-4"><a href="bioconductor.html#cb69-4"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;IRanges&quot;</span>)</span>
<span id="cb69-5"><a href="bioconductor.html#cb69-5"></a><span class="co">## .libPaths()</span></span></code></pre></div>
<p><strong>Learning and support</strong></p>
<p>Learn &gt;&gt; Support site</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="bioconductor.html#cb70-1"></a><span class="kw">library</span>(IRanges)</span>
<span id="cb70-2"><a href="bioconductor.html#cb70-2"></a></span>
<span id="cb70-3"><a href="bioconductor.html#cb70-3"></a><span class="kw">vignette</span>(<span class="dt">package=</span><span class="st">&quot;IRanges&quot;</span>)</span>
<span id="cb70-4"><a href="bioconductor.html#cb70-4"></a><span class="kw">browseVignettes</span>(<span class="st">&quot;IRanges&quot;</span>)</span>
<span id="cb70-5"><a href="bioconductor.html#cb70-5"></a><span class="kw">vignette</span>(<span class="st">&quot;IRangesOverview&quot;</span>, <span class="dt">package=</span><span class="st">&quot;IRanges&quot;</span>)</span>
<span id="cb70-6"><a href="bioconductor.html#cb70-6"></a></span>
<span id="cb70-7"><a href="bioconductor.html#cb70-7"></a>ir1 &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">width=</span><span class="dv">10</span><span class="op">:</span><span class="dv">1</span>)</span>
<span id="cb70-8"><a href="bioconductor.html#cb70-8"></a>ir1</span>
<span id="cb70-9"><a href="bioconductor.html#cb70-9"></a><span class="kw">class</span>(ir1)</span>
<span id="cb70-10"><a href="bioconductor.html#cb70-10"></a><span class="kw">methods</span>(<span class="dt">class=</span><span class="st">&quot;IRanges&quot;</span>)</span>
<span id="cb70-11"><a href="bioconductor.html#cb70-11"></a></span>
<span id="cb70-12"><a href="bioconductor.html#cb70-12"></a><span class="kw">example</span>(IRanges)</span>
<span id="cb70-13"><a href="bioconductor.html#cb70-13"></a>?IRanges</span>
<span id="cb70-14"><a href="bioconductor.html#cb70-14"></a>??IRanges</span></code></pre></div>
</div>
<div id="data-structure-and-management-for-genome-scale-experiments" class="section level2">
<h2><span class="header-section-number">4.3</span> Data structure and management for genome-scale experiments</h2>
<p>연구를 수행하면서 잘못된 데이터 관리로 인한 손실을 줄이기 위해서는 고도로 효율적이고 신뢰성 높은 툴이 필요합니다. 일반적으로 사용되는 엑셀 기반의 데이터 관리는 신뢰성과 파이프라인 측면에서 적합하지 않습니다. 즉, 특정 파이프라인을 만들어 놓고 이에 따라서 일관적으로 데이터를 관리할 때 높은 신뢰성의 데이터를 확보할 수 있고 과학적 추론의 바탕이 될 수 있습니다. 따라서 복잡한 데이터의 구조를 체계적으로 정의하고 정형화된 도구들로 분석 파이프라인을 개발할 필요가 있습니다.</p>
<div id="expressionset-class" class="section level3">
<h3><span class="header-section-number">4.3.1</span> ExpressionSet Class</h3>
<p><code>Biobase</code> 패키지는 지놈 데이터를 관리하기 위한 표준화된 데이터 구조 class인 <code>ExpressionSet</code>를 제공합니다.</p>
<p><img src="images/04/expressionset.JPG" /></p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="bioconductor.html#cb71-1"></a><span class="kw">library</span>(Biobase)</span>
<span id="cb71-2"><a href="bioconductor.html#cb71-2"></a>?ExpressionSet</span>
<span id="cb71-3"><a href="bioconductor.html#cb71-3"></a></span>
<span id="cb71-4"><a href="bioconductor.html#cb71-4"></a><span class="kw">data</span>(sample.ExpressionSet)</span>
<span id="cb71-5"><a href="bioconductor.html#cb71-5"></a>sample.ExpressionSet</span>
<span id="cb71-6"><a href="bioconductor.html#cb71-6"></a><span class="co">#browseVignettes(&quot;Biobase&quot;)</span></span>
<span id="cb71-7"><a href="bioconductor.html#cb71-7"></a><span class="co">#methods(class=&quot;ExpressionSet&quot;)</span></span>
<span id="cb71-8"><a href="bioconductor.html#cb71-8"></a><span class="kw">featureNames</span>(sample.ExpressionSet)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb71-9"><a href="bioconductor.html#cb71-9"></a><span class="kw">sampleNames</span>(sample.ExpressionSet)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb71-10"><a href="bioconductor.html#cb71-10"></a></span>
<span id="cb71-11"><a href="bioconductor.html#cb71-11"></a>expression_data &lt;-<span class="st"> </span><span class="kw">exprs</span>(sample.ExpressionSet)</span>
<span id="cb71-12"><a href="bioconductor.html#cb71-12"></a><span class="kw">dim</span>(expression_data)</span>
<span id="cb71-13"><a href="bioconductor.html#cb71-13"></a>feature_data &lt;-<span class="st"> </span><span class="kw">fData</span>(sample.ExpressionSet)</span>
<span id="cb71-14"><a href="bioconductor.html#cb71-14"></a><span class="kw">dim</span>(feature_data)</span>
<span id="cb71-15"><a href="bioconductor.html#cb71-15"></a>pheno_data &lt;-<span class="st"> </span><span class="kw">pData</span>(sample.ExpressionSet)</span>
<span id="cb71-16"><a href="bioconductor.html#cb71-16"></a><span class="kw">dim</span>(pheno_data)</span></code></pre></div>
<p><img src="images/bioconductor_strucure.jpg" /></p>
<p>급성 림프구성 백혈병 데이터</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="bioconductor.html#cb72-1"></a><span class="kw">library</span>(ALL)</span>
<span id="cb72-2"><a href="bioconductor.html#cb72-2"></a><span class="kw">data</span>(ALL)</span>
<span id="cb72-3"><a href="bioconductor.html#cb72-3"></a><span class="kw">class</span>(ALL)</span>
<span id="cb72-4"><a href="bioconductor.html#cb72-4"></a>ALL</span>
<span id="cb72-5"><a href="bioconductor.html#cb72-5"></a><span class="kw">browseVignettes</span>(<span class="st">&quot;ALL&quot;</span>)</span>
<span id="cb72-6"><a href="bioconductor.html#cb72-6"></a><span class="co">#methods(class=class(ALL))</span></span>
<span id="cb72-7"><a href="bioconductor.html#cb72-7"></a><span class="kw">pData</span>(ALL)</span>
<span id="cb72-8"><a href="bioconductor.html#cb72-8"></a><span class="kw">pData</span>(ALL)<span class="op">$</span>age</span>
<span id="cb72-9"><a href="bioconductor.html#cb72-9"></a><span class="kw">fData</span>(ALL)</span>
<span id="cb72-10"><a href="bioconductor.html#cb72-10"></a><span class="co">## </span></span>
<span id="cb72-11"><a href="bioconductor.html#cb72-11"></a><span class="kw">phenoData</span>(ALL)</span>
<span id="cb72-12"><a href="bioconductor.html#cb72-12"></a><span class="kw">varMetadata</span>(<span class="kw">phenoData</span>(ALL))</span>
<span id="cb72-13"><a href="bioconductor.html#cb72-13"></a><span class="kw">featureData</span>(ALL)</span></code></pre></div>
<p>메타데이터는 데이터(data)에 대한 데이터로서 일반적으로는 어떤 목적을 가지고 만들어진 데이터로 정의됩니다.</p>
<p><img src="images/geometa.JPG" /></p>
<p>Gene expression omnibus (GEO)
GEO는 microarray, next-generation sequencing 등의 high-throughput 유전체 데이터를 보유한 공공 저장소입니다.</p>
<ul>
<li>대규모 기능유전체 데이터베이스</li>
<li>데이터 기탁 쉽게 만들고 고수준 QC 유지</li>
<li>사용하기 쉬운 인터페이스 유지</li>
</ul>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/info/overview.html">GEO</a></p>
<p><img src="images/04/geo_overview.jpg" /></p>
<!-- [GEO accession data GSE62944](https://www.bioconductor.org/packages/release/data/experiment/html/GSE62944.html) 설명  -->
<!-- ```{r, eval=F} -->
<!-- if (!requireNamespace("BiocManager", quietly = TRUE)) -->
<!--     install.packages("BiocManager") -->
<!-- BiocManager::install("GSE62944") -->
<!-- library(GSE62944) -->
<!-- browseVignettes("GSE62944") -->
<!-- BiocManager::install("ExperimentHub") -->
<!-- browseVignettes("ExperimentHub") -->
<!-- library(ExperimentHub) -->
<!-- eh = ExperimentHub(localHub=FALSE) -->
<!-- query(eh, "GSE62944") -->
<!-- tcga_data <- eh[["EH1043"]] -->
<!-- ``` -->
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="bioconductor.html#cb73-1"></a><span class="kw">library</span>(GEOquery)</span>
<span id="cb73-2"><a href="bioconductor.html#cb73-2"></a><span class="kw">browseVignettes</span>(<span class="st">&quot;GEOquery&quot;</span>)</span></code></pre></div>
<p>The <code>GDS</code> class</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="bioconductor.html#cb74-1"></a>gds &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">&quot;extdata/GDS507.soft.gz&quot;</span>,<span class="dt">package=</span><span class="st">&quot;GEOquery&quot;</span>))</span>
<span id="cb74-2"><a href="bioconductor.html#cb74-2"></a><span class="kw">class</span>(gds)</span>
<span id="cb74-3"><a href="bioconductor.html#cb74-3"></a><span class="kw">methods</span>(<span class="dt">class=</span><span class="kw">class</span>(gds))</span>
<span id="cb74-4"><a href="bioconductor.html#cb74-4"></a><span class="kw">Table</span>(gds)</span>
<span id="cb74-5"><a href="bioconductor.html#cb74-5"></a><span class="kw">dataTable</span>(gds)</span>
<span id="cb74-6"><a href="bioconductor.html#cb74-6"></a><span class="kw">show</span>(gds)</span>
<span id="cb74-7"><a href="bioconductor.html#cb74-7"></a><span class="kw">Columns</span>(gds)</span></code></pre></div>
<p>The <code>GSM</code> class</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="bioconductor.html#cb75-1"></a>gsm &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">&quot;extdata/GSM11805.txt.gz&quot;</span>,<span class="dt">package=</span><span class="st">&quot;GEOquery&quot;</span>))</span>
<span id="cb75-2"><a href="bioconductor.html#cb75-2"></a><span class="kw">methods</span>(<span class="dt">class=</span><span class="kw">class</span>(gsm))</span>
<span id="cb75-3"><a href="bioconductor.html#cb75-3"></a><span class="kw">head</span>(<span class="kw">Meta</span>(gsm))</span>
<span id="cb75-4"><a href="bioconductor.html#cb75-4"></a><span class="kw">Table</span>(gsm)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</span>
<span id="cb75-5"><a href="bioconductor.html#cb75-5"></a><span class="kw">dim</span>(<span class="kw">Table</span>(gsm))</span>
<span id="cb75-6"><a href="bioconductor.html#cb75-6"></a><span class="kw">Columns</span>(gsm)</span></code></pre></div>
<p>The <code>GSE</code> class</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="bioconductor.html#cb76-1"></a>gse &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">&quot;extdata/GSE781_family.soft.gz&quot;</span>,<span class="dt">package=</span><span class="st">&quot;GEOquery&quot;</span>))</span>
<span id="cb76-2"><a href="bioconductor.html#cb76-2"></a><span class="kw">methods</span>(<span class="dt">class=</span><span class="kw">class</span>(gse))</span>
<span id="cb76-3"><a href="bioconductor.html#cb76-3"></a><span class="kw">head</span>(<span class="kw">Meta</span>(gse))</span>
<span id="cb76-4"><a href="bioconductor.html#cb76-4"></a><span class="kw">names</span>(<span class="kw">GSMList</span>(gse))</span>
<span id="cb76-5"><a href="bioconductor.html#cb76-5"></a><span class="kw">class</span>(<span class="kw">GSMList</span>(gse)[[<span class="dv">1</span>]])</span></code></pre></div>
<p>ExpressionSet class의 GES 데이터 받기 <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2553">GSE2553</a></p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="bioconductor.html#cb77-1"></a>gse2553 &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="st">&#39;GSE2553&#39;</span>,<span class="dt">GSEMatrix=</span><span class="ot">TRUE</span>)</span>
<span id="cb77-2"><a href="bioconductor.html#cb77-2"></a>gse2553</span>
<span id="cb77-3"><a href="bioconductor.html#cb77-3"></a><span class="kw">class</span>(gse2553)</span>
<span id="cb77-4"><a href="bioconductor.html#cb77-4"></a><span class="kw">class</span>(gse2553[[<span class="dv">1</span>]])</span>
<span id="cb77-5"><a href="bioconductor.html#cb77-5"></a>mygse &lt;-<span class="st"> </span>gse2553[[<span class="dv">1</span>]]</span>
<span id="cb77-6"><a href="bioconductor.html#cb77-6"></a><span class="kw">pData</span>(mygse)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<p>GDS 데이터를 ExpressionSet class로 변환하기</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="bioconductor.html#cb78-1"></a>gds &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="dt">filename=</span><span class="kw">system.file</span>(<span class="st">&quot;extdata/GDS507.soft.gz&quot;</span>,<span class="dt">package=</span><span class="st">&quot;GEOquery&quot;</span>))</span>
<span id="cb78-2"><a href="bioconductor.html#cb78-2"></a><span class="kw">class</span>(gds)</span>
<span id="cb78-3"><a href="bioconductor.html#cb78-3"></a>eset &lt;-<span class="st"> </span><span class="kw">GDS2eSet</span>(gds, <span class="dt">do.log2=</span><span class="ot">TRUE</span>)</span>
<span id="cb78-4"><a href="bioconductor.html#cb78-4"></a>eset</span>
<span id="cb78-5"><a href="bioconductor.html#cb78-5"></a></span>
<span id="cb78-6"><a href="bioconductor.html#cb78-6"></a><span class="co">## boxplot/vioplot for the 17 samples</span></span></code></pre></div>
<p>급성 림프구성 백혈병 데이터 (Annotation)</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="bioconductor.html#cb79-1"></a><span class="kw">library</span>(ALL)</span>
<span id="cb79-2"><a href="bioconductor.html#cb79-2"></a><span class="kw">data</span>(ALL)</span>
<span id="cb79-3"><a href="bioconductor.html#cb79-3"></a>ALL</span>
<span id="cb79-4"><a href="bioconductor.html#cb79-4"></a><span class="kw">featureData</span>(ALL)</span>
<span id="cb79-5"><a href="bioconductor.html#cb79-5"></a></span>
<span id="cb79-6"><a href="bioconductor.html#cb79-6"></a><span class="co">## </span></span>
<span id="cb79-7"><a href="bioconductor.html#cb79-7"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb79-8"><a href="bioconductor.html#cb79-8"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb79-9"><a href="bioconductor.html#cb79-9"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;hgu95av2.db&quot;</span>)</span>
<span id="cb79-10"><a href="bioconductor.html#cb79-10"></a><span class="kw">library</span>(hgu95av2.db)</span>
<span id="cb79-11"><a href="bioconductor.html#cb79-11"></a></span>
<span id="cb79-12"><a href="bioconductor.html#cb79-12"></a><span class="kw">browseVignettes</span>(<span class="st">&quot;hgu95av2.db&quot;</span>)</span>
<span id="cb79-13"><a href="bioconductor.html#cb79-13"></a><span class="kw">help</span>(<span class="dt">package=</span><span class="st">&quot;hgu95av2.db&quot;</span>)</span>
<span id="cb79-14"><a href="bioconductor.html#cb79-14"></a></span>
<span id="cb79-15"><a href="bioconductor.html#cb79-15"></a><span class="kw">featureNames</span>(ALL)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb79-16"><a href="bioconductor.html#cb79-16"></a>ids &lt;-<span class="st"> </span><span class="kw">featureNames</span>(ALL)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb79-17"><a href="bioconductor.html#cb79-17"></a><span class="kw">as.list</span>(hgu95av2ENTREZID[ids])</span></code></pre></div>
</div>
</div>
<div id="biostrings-package-for-sequence-analysis" class="section level2">
<h2><span class="header-section-number">4.4</span> Biostrings package for sequence analysis</h2>
<p>먼저 <code>Biostrings</code>는 아래와 같이 Bioconductor에서 제공하는 코드를 이용해서 설치할 수 있습니다.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="bioconductor.html#cb80-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb80-2"><a href="bioconductor.html#cb80-2"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb80-3"><a href="bioconductor.html#cb80-3"></a></span>
<span id="cb80-4"><a href="bioconductor.html#cb80-4"></a>BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;Biostrings&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="bioconductor.html#cb81-1"></a><span class="kw">library</span>(Biostrings)</span></code></pre></div>
<p>Biostrings 패키지는 기본적으로 <code>XString, XStringSet, XStringViews</code> 3가지의 class를 정의하고 있습니다. XString은 DNA나 RNA, AA 등 생물학적 서열 한 가닥을 다루기위한 클래스이며 XStringSet은 여러 가닥을 다루기위한 클래스 입니다.</p>
<p><img src="images/04/biostrings.JPG" /></p>
<p>다음 내장변수 들은 <code>Biostrings</code> 패키지를 로드하면 자동으로 저장되는 변수들로 생물학적 서열을 미리 정의해 놓았습니다.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="bioconductor.html#cb82-1"></a>DNA_BASES</span>
<span id="cb82-2"><a href="bioconductor.html#cb82-2"></a>DNA_ALPHABET</span>
<span id="cb82-3"><a href="bioconductor.html#cb82-3"></a>IUPAC_CODE_MAP</span></code></pre></div>
<p>위 변수들을 이용하면 다음처럼 <code>sample()</code> 함수를 이용해서 랜덤하게 DNA 서열을 얻을 수 있습니다. <code>DNA_BASES</code>가 4개 길이를 갖는 벡터인데 이 중 10개를 뽑으려면 <code>replace=T</code>로 해야 합니다.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="bioconductor.html#cb83-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">sample</span>(DNA_BASES, <span class="dv">10</span>, <span class="dt">replace =</span> T)</span>
<span id="cb83-2"><a href="bioconductor.html#cb83-2"></a>x0</span>
<span id="cb83-3"><a href="bioconductor.html#cb83-3"></a>x1 &lt;-<span class="st"> </span><span class="kw">paste</span>(x0, <span class="dt">collapse=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb83-4"><a href="bioconductor.html#cb83-4"></a>x1</span></code></pre></div>
<p>XString 클래스 특히 DNAString 클래스의 사용법은 다음 코드를 통해 익힐 수 있습니다. <code>length</code> 함수는 핵산의 갯수를 (<code>DNAStringSet</code> 타입의 변수에서 length는 DNA 가닥의 갯수이며 핵산의 갯수는 <code>nchar</code>함수로 얻어낼 수 있습니다. <code>toString</code>은 DNAString 타입을 단순 문자열로 변환해주는 함수이며 상보서열, 역상보서열 등의 정보도 <code>complement</code>, <code>reverseComplement</code> 등을 사용하여 찾아낼 수 있습니다.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="bioconductor.html#cb84-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">sample</span>(DNA_BASES, <span class="dv">10</span>, <span class="dt">replace =</span> T), <span class="dt">collapse=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb84-2"><a href="bioconductor.html#cb84-2"></a>x1 =<span class="st"> </span><span class="kw">DNAString</span>(x0)</span>
<span id="cb84-3"><a href="bioconductor.html#cb84-3"></a><span class="kw">class</span>(x0)</span>
<span id="cb84-4"><a href="bioconductor.html#cb84-4"></a><span class="kw">class</span>(x1)</span>
<span id="cb84-5"><a href="bioconductor.html#cb84-5"></a><span class="kw">length</span>(x1)</span>
<span id="cb84-6"><a href="bioconductor.html#cb84-6"></a><span class="kw">toString</span>(x1)</span>
<span id="cb84-7"><a href="bioconductor.html#cb84-7"></a><span class="kw">complement</span>(x1)</span>
<span id="cb84-8"><a href="bioconductor.html#cb84-8"></a>Biostrings<span class="op">::</span><span class="kw">complement</span>(x1)</span>
<span id="cb84-9"><a href="bioconductor.html#cb84-9"></a><span class="kw">reverseComplement</span>(x1)</span></code></pre></div>
<p>DNAString의 인덱싱은 vector (string)과 같으며 DNAStringSet은 list의 인덱싱과 같습니다.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="bioconductor.html#cb85-1"></a><span class="co">## indexing</span></span>
<span id="cb85-2"><a href="bioconductor.html#cb85-2"></a>x1[<span class="dv">1</span>]</span>
<span id="cb85-3"><a href="bioconductor.html#cb85-3"></a>x1[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span>
<span id="cb85-4"><a href="bioconductor.html#cb85-4"></a><span class="kw">subseq</span>(x1, <span class="dt">start=</span><span class="dv">3</span>, <span class="dt">end=</span><span class="dv">5</span>)</span>
<span id="cb85-5"><a href="bioconductor.html#cb85-5"></a><span class="kw">subseq</span>(x1, <span class="dv">3</span>, <span class="dv">5</span>)</span>
<span id="cb85-6"><a href="bioconductor.html#cb85-6"></a></span>
<span id="cb85-7"><a href="bioconductor.html#cb85-7"></a><span class="co">## letter frequency</span></span>
<span id="cb85-8"><a href="bioconductor.html#cb85-8"></a><span class="kw">alphabetFrequency</span>(x1, <span class="dt">baseOnly=</span><span class="ot">TRUE</span>, <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span>
<span id="cb85-9"><a href="bioconductor.html#cb85-9"></a><span class="kw">letterFrequency</span>(x1, <span class="kw">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<p><strong>Exercise</strong></p>
<p>시작코돈과 종결코돈이 있는 길이 36bp 짜리 DNA (랜덤) 서열을 만들고 GC 비율을 계산하시오</p>
<p><code>DNAStringSet</code> class는 여러개의 <code>DNAString</code> 을 모아 놓은 집합이라고 보면 됩니다. <code>length</code> 함수는 DNA string의 갯수이며 <code>width</code> 또는 <code>nchar</code> 함수로 각 string의 길이를 구할 수 있으며 이 외 대부분의 <code>DNAString</code> 에서 사용되는 함수가 동일하게 사용될 수 있습니다.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="bioconductor.html#cb86-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CTC-NACCAGTAT&quot;</span>, <span class="st">&quot;TTGA&quot;</span>, <span class="st">&quot;TACCTAGAG&quot;</span>)</span>
<span id="cb86-2"><a href="bioconductor.html#cb86-2"></a>x1 &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(x0)</span>
<span id="cb86-3"><a href="bioconductor.html#cb86-3"></a><span class="kw">class</span>(x0)</span>
<span id="cb86-4"><a href="bioconductor.html#cb86-4"></a><span class="kw">class</span>(x1)</span>
<span id="cb86-5"><a href="bioconductor.html#cb86-5"></a><span class="kw">names</span>(x1)</span>
<span id="cb86-6"><a href="bioconductor.html#cb86-6"></a><span class="kw">names</span>(x1) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>)</span>
<span id="cb86-7"><a href="bioconductor.html#cb86-7"></a><span class="kw">length</span>(x1)</span>
<span id="cb86-8"><a href="bioconductor.html#cb86-8"></a><span class="kw">width</span>(x1)</span>
<span id="cb86-9"><a href="bioconductor.html#cb86-9"></a><span class="kw">subseq</span>(x1, <span class="dv">2</span>, <span class="dv">4</span>)</span>
<span id="cb86-10"><a href="bioconductor.html#cb86-10"></a>x1[[<span class="dv">1</span>]]</span>
<span id="cb86-11"><a href="bioconductor.html#cb86-11"></a>x1[<span class="dv">1</span>]</span>
<span id="cb86-12"><a href="bioconductor.html#cb86-12"></a></span>
<span id="cb86-13"><a href="bioconductor.html#cb86-13"></a></span>
<span id="cb86-14"><a href="bioconductor.html#cb86-14"></a>x3 &lt;-<span class="st"> </span><span class="kw">DNAString</span>(<span class="st">&quot;ATGAGTAGTTAG&quot;</span>)</span>
<span id="cb86-15"><a href="bioconductor.html#cb86-15"></a>x4 &lt;-<span class="st"> </span><span class="kw">c</span>(x1, <span class="kw">DNAStringSet</span>(x3))</span>
<span id="cb86-16"><a href="bioconductor.html#cb86-16"></a>x4[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb86-17"><a href="bioconductor.html#cb86-17"></a>x4</span>
<span id="cb86-18"><a href="bioconductor.html#cb86-18"></a><span class="kw">alphabetFrequency</span>(x1, <span class="dt">baseOnly=</span><span class="ot">TRUE</span>, <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span>
<span id="cb86-19"><a href="bioconductor.html#cb86-19"></a><span class="kw">letterFrequency</span>(x1, <span class="kw">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span>
<span id="cb86-20"><a href="bioconductor.html#cb86-20"></a><span class="kw">rowSums</span>(<span class="kw">letterFrequency</span>(x1, <span class="kw">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="dt">as.prob=</span><span class="ot">TRUE</span>))</span>
<span id="cb86-21"><a href="bioconductor.html#cb86-21"></a><span class="kw">subseq</span>(x4, <span class="dv">2</span>, <span class="dv">4</span>)</span></code></pre></div>
<p>아래는 가장 직관적으로 생각할 수 있는 <code>for</code>를 이용한 방법입니다. 즉, 10개 저장소를 갖는 x0 변수를 미리 생성해 두고 for 문을 돌면서 서열을 하나씩 만들어 저장하는 방법입니다.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="bioconductor.html#cb87-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;&quot;</span>, <span class="dv">10</span>)</span>
<span id="cb87-2"><a href="bioconductor.html#cb87-2"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(x0)){</span>
<span id="cb87-3"><a href="bioconductor.html#cb87-3"></a>  tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">sample</span>(DNA_BASES, <span class="dv">30</span>, <span class="dt">replace =</span> T), <span class="dt">collapse=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb87-4"><a href="bioconductor.html#cb87-4"></a>  x0[i] &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;ATG&quot;</span>, tmp, <span class="st">&quot;TAG&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb87-5"><a href="bioconductor.html#cb87-5"></a>}</span>
<span id="cb87-6"><a href="bioconductor.html#cb87-6"></a>x0</span></code></pre></div>
<p>위 코드를 함수로 만들어 보겠습니다. random dna를 만들 때 길이만 다를뿐 같은 코드를 반복해서 사용하고 있습니다. 이럴 경우 DNA 길이를 사용자가 정해주도록 input parameter로 하고 해당 파라메터를 받아 DNA를 만들어 주는 함수를 만들어 사용하면 편리합니다.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="bioconductor.html#cb88-1"></a>random_dna &lt;-<span class="st"> </span><span class="cf">function</span>(len){</span>
<span id="cb88-2"><a href="bioconductor.html#cb88-2"></a>  tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">sample</span>(DNA_BASES, len, <span class="dt">replace =</span> T), <span class="dt">collapse=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb88-3"><a href="bioconductor.html#cb88-3"></a>  x0 &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;ATG&quot;</span>, tmp, <span class="st">&quot;TAG&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb88-4"><a href="bioconductor.html#cb88-4"></a>  <span class="kw">return</span>(x0)</span>
<span id="cb88-5"><a href="bioconductor.html#cb88-5"></a>}</span>
<span id="cb88-6"><a href="bioconductor.html#cb88-6"></a><span class="kw">random_dna</span>(<span class="dt">len=</span><span class="dv">30</span>)</span>
<span id="cb88-7"><a href="bioconductor.html#cb88-7"></a><span class="kw">random_dna</span>(<span class="dt">len=</span><span class="dv">40</span>)</span></code></pre></div>
<p>파라메터로 넘겨진 len 값이 sample 함수의 len에 사용된 것을 참고하세요.</p>
<p>이제 길이 30bp짜리 10개의 서열을 반복해서 만들 때 위 함수를 앞서와 같이 for문을 이용하여 10번 반복해서 실행해 주면 같은 결과를 얻습니다. 위와 같이 함수를 만들어 두면 언제든 DNA 서열을 만들 때 재사용 할 수 있습니다.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="bioconductor.html#cb89-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;&quot;</span>, <span class="dv">10</span>)</span>
<span id="cb89-2"><a href="bioconductor.html#cb89-2"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(x0)){</span>
<span id="cb89-3"><a href="bioconductor.html#cb89-3"></a>  x0[i] &lt;-<span class="st"> </span><span class="kw">random_dna</span>(<span class="dv">30</span>)</span>
<span id="cb89-4"><a href="bioconductor.html#cb89-4"></a>}</span>
<span id="cb89-5"><a href="bioconductor.html#cb89-5"></a>x0</span></code></pre></div>
<p>그런데 R에는 <code>apply</code> 와 같은 행렬연산 함수가 있어서 for문을 사용하지 않고 편리하게 반복문을 실행할 수 있습니다. <code>replicate</code> 함수는 <code>apply</code>와 같은 기능으로 list나 vector 변수에 대해서 사용할 수 있습니다. 즉, 다음과 같이 사용자가 원하는 함수를 반복해서 실행하고 반복 수 만큼의 길이를 갖는 결과를 반환합니다.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="bioconductor.html#cb90-1"></a>x0 &lt;-<span class="st"> </span><span class="kw">replicate</span>(<span class="dv">10</span>, <span class="kw">random_dna</span>(<span class="dv">30</span>))</span>
<span id="cb90-2"><a href="bioconductor.html#cb90-2"></a>x0</span>
<span id="cb90-3"><a href="bioconductor.html#cb90-3"></a>x1 &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(x0)</span>
<span id="cb90-4"><a href="bioconductor.html#cb90-4"></a>x1</span></code></pre></div>
<p>위 x0 스트링들을 XStringSet으로 바꾸고 GC 비율을 구한 후 bargraph를 그리겠습니다. gc_ratio가 G와 C의 비율값을 저장한 10x2 테이블이므로 x축에 10개의 서열과 각 서열의 GC비율을 나타내고 y축에 비율 값을 그리는 것으로 생각한 후 ggplot의 aes와 파라메터를 적절히 지정해 줍니다.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="bioconductor.html#cb91-1"></a>x1 &lt;-<span class="st"> </span><span class="kw">DNAStringSet</span>(x0)</span>
<span id="cb91-2"><a href="bioconductor.html#cb91-2"></a>gc_ratio1 &lt;-<span class="st"> </span><span class="kw">letterFrequency</span>(x1, <span class="kw">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="dt">as.prob=</span><span class="ot">TRUE</span>)</span>
<span id="cb91-3"><a href="bioconductor.html#cb91-3"></a>gc_ratio2 &lt;-<span class="st"> </span><span class="kw">rowSums</span>(gc_ratio1)</span>
<span id="cb91-4"><a href="bioconductor.html#cb91-4"></a><span class="kw">barplot</span>(gc_ratio2, <span class="dt">beside=</span>T)</span></code></pre></div>
<p><code>Biostrings</code>의 또 다른 class인 <code>XStringView</code>는 <code>XString</code> class의 DNA서열을 사용자가 원하는대로 볼 수 있는 인터페이스를 제공합니다. 사용법은 다음과 같습니다.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="bioconductor.html#cb92-1"></a>x2 &lt;-<span class="st"> </span>x1[[<span class="dv">1</span>]]</span>
<span id="cb92-2"><a href="bioconductor.html#cb92-2"></a><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="dv">1</span>, <span class="dt">width=</span><span class="dv">20</span>)</span>
<span id="cb92-3"><a href="bioconductor.html#cb92-3"></a><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="dv">1</span>, <span class="dt">end=</span><span class="dv">4</span>)</span>
<span id="cb92-4"><a href="bioconductor.html#cb92-4"></a><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="dt">end=</span><span class="dv">4</span>)</span>
<span id="cb92-5"><a href="bioconductor.html#cb92-5"></a><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>), <span class="dt">width=</span><span class="dv">20</span>)</span>
<span id="cb92-6"><a href="bioconductor.html#cb92-6"></a><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>), <span class="dt">width=</span><span class="dv">20</span>)</span>
<span id="cb92-7"><a href="bioconductor.html#cb92-7"></a>i &lt;-<span class="st"> </span><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>), <span class="dt">width=</span><span class="dv">20</span>)</span></code></pre></div>
<p>다음과 같이 한 서열에 대한 여러 부분의 서열 조각도 볼 수 있으며 <code>gaps</code> 함수는 매개변수로 주어진 서열 view의 구간을 제외한 나머지 구간의 서열을 보여주는 함수입니다. <code>successiveviews</code> 함수는 처음 서열부터 매개변수 width에 주어진 갯수 만큼의 서열을 보여주며 <code>rep()</code> 함수를 이용해서 서열의 처음부터 끝까지 보여주는 기능을 합니다.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="bioconductor.html#cb93-1"></a>v &lt;-<span class="st"> </span><span class="kw">Views</span>(x2, <span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">10</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">15</span>))</span>
<span id="cb93-2"><a href="bioconductor.html#cb93-2"></a><span class="kw">gaps</span>(v)</span>
<span id="cb93-3"><a href="bioconductor.html#cb93-3"></a></span>
<span id="cb93-4"><a href="bioconductor.html#cb93-4"></a><span class="kw">successiveViews</span>(x2, <span class="dt">width=</span><span class="dv">20</span>)</span>
<span id="cb93-5"><a href="bioconductor.html#cb93-5"></a><span class="kw">successiveViews</span>(x2, <span class="dt">width=</span><span class="kw">rep</span>(<span class="dv">20</span>, <span class="dv">2</span>))</span>
<span id="cb93-6"><a href="bioconductor.html#cb93-6"></a><span class="kw">successiveViews</span>(x2, <span class="dt">width=</span><span class="kw">rep</span>(<span class="dv">20</span>, <span class="dv">3</span>))</span></code></pre></div>
<p>서열 읽고 쓰기</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="bioconductor.html#cb94-1"></a><span class="kw">writeXStringSet</span>(<span class="kw">DNAStringSet</span>(x0), <span class="st">&quot;myfastaseq.fasta&quot;</span>, <span class="dt">format=</span><span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb94-2"><a href="bioconductor.html#cb94-2"></a></span>
<span id="cb94-3"><a href="bioconductor.html#cb94-3"></a>myseq &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(<span class="st">&quot;myfastaseq.fasta&quot;</span>, <span class="dt">format=</span><span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb94-4"><a href="bioconductor.html#cb94-4"></a>myseq</span></code></pre></div>
<p><strong>Exercise</strong></p>
<p>1000bp 길이의 랜덤 DNA 서열을 만들고 40bp 단위의 길이로 보는 코드를 만들어 보겠습니다. 앞서 만들어둔 <code>random_dna()</code> 함수를 사용하면 되며 <code>successiveview</code> 함수를 사용해야 하므로 <code>DNAString</code>으로 변환이 필요하며 서열의 길이에 따라서 <code>rep()</code> 를 이용하여 반복 횟수를 자동 계산합니다.</p>
</div>
<div id="download-sequence-information-from-ncbi" class="section level2">
<h2><span class="header-section-number">4.5</span> Download sequence information from NCBI</h2>
<p>전세계 연구자들이 서열 데이터를 분석하는데 가장 많이 이용하는 사이트 중 하나가 NCBI 이며 따라서 NCBI에서는 연구자들이 데이터베이스에 접근하기위한 편리한 방법을 제공하고 있고 그 중 하나가 Entrez 입니다.</p>
<p>R에서도 Entrez 기능을 도입한 package들이 제공되고 있으며 그 중 하나가 <code>rentrez</code> 입니다. <a href="https://www.ncbi.nlm.nih.gov/books/NBK25500/" class="uri">https://www.ncbi.nlm.nih.gov/books/NBK25500/</a> 이 곳의 Downloading Full Records 를 참고하시면 좋습니다. Entrez는 대략적으로 다음 9개의 유틸리티를 제공합니다.</p>
<blockquote>
<p>EInfo (database statistics)<br />
ESearch (text searches)<br />
EPost (UID uploads)<br />
ESummary (document summary downloads)<br />
EFetch (data record downloads)<br />
ELink (Entrez links)<br />
EGQuery (global query)<br />
ESpell (spelling suggestions)<br />
ECitMatch (batch citation searching in PubMed)</p>
</blockquote>
<p>이 중 <code>ESerach</code>, <code>EPost</code>, <code>ESummary</code>, <code>EFetch</code> 등이 많이 사용하는 유틸이며 정보를 다운로드 받을 경우는 <code>EFetch</code> 를 주로 사용하게 됩니다. 예제로 뎅기바이러스 서열 4종을 다운로드 하겠습니다. NCBI의 accession 번호를 알 경우이며 각각에 대한 accession no.는 NC_001477, NC_001474, NC_001475 and NC_002640 입니다.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="bioconductor.html#cb95-1"></a><span class="kw">library</span>(rentrez)</span>
<span id="cb95-2"><a href="bioconductor.html#cb95-2"></a><span class="kw">entrez_dbs</span>()</span>
<span id="cb95-3"><a href="bioconductor.html#cb95-3"></a><span class="kw">entrez_db_summary</span>(<span class="st">&quot;nuccore&quot;</span>)</span>
<span id="cb95-4"><a href="bioconductor.html#cb95-4"></a></span>
<span id="cb95-5"><a href="bioconductor.html#cb95-5"></a><span class="co">##</span></span>
<span id="cb95-6"><a href="bioconductor.html#cb95-6"></a>acc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NC_001477&quot;</span>, <span class="st">&quot;NC_001474&quot;</span>, <span class="st">&quot;NC_001475&quot;</span>, <span class="st">&quot;NC_002640&quot;</span>)</span>
<span id="cb95-7"><a href="bioconductor.html#cb95-7"></a>all_recs &lt;-<span class="st"> </span><span class="kw">entrez_fetch</span>(<span class="dt">db=</span><span class="st">&quot;nuccore&quot;</span>, <span class="dt">id=</span>acc[<span class="dv">1</span>], <span class="dt">rettype=</span><span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb95-8"><a href="bioconductor.html#cb95-8"></a><span class="kw">cat</span>(<span class="kw">strwrap</span>(<span class="kw">substr</span>(all_recs, <span class="dv">1</span>, <span class="dv">500</span>)), <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</span>
<span id="cb95-9"><a href="bioconductor.html#cb95-9"></a><span class="kw">write</span>(all_recs, <span class="dt">file=</span><span class="st">&quot;myseq.fasta&quot;</span>)</span>
<span id="cb95-10"><a href="bioconductor.html#cb95-10"></a>dang &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(<span class="st">&quot;myseq.fasta&quot;</span>, <span class="dt">format=</span><span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb95-11"><a href="bioconductor.html#cb95-11"></a></span>
<span id="cb95-12"><a href="bioconductor.html#cb95-12"></a><span class="co">##</span></span>
<span id="cb95-13"><a href="bioconductor.html#cb95-13"></a>r_search &lt;-<span class="st"> </span><span class="kw">entrez_search</span>(<span class="dt">db=</span><span class="st">&quot;pubmed&quot;</span>, <span class="dt">term=</span><span class="st">&quot;R Language&quot;</span>)</span>
<span id="cb95-14"><a href="bioconductor.html#cb95-14"></a>all_recs &lt;-<span class="st"> </span><span class="kw">entrez_fetch</span>(<span class="dt">db=</span><span class="st">&quot;pubmed&quot;</span>, <span class="dt">id=</span>r_search<span class="op">$</span>ids, <span class="dt">rettype=</span><span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb95-15"><a href="bioconductor.html#cb95-15"></a><span class="kw">write</span>(all_recs, <span class="dt">file=</span><span class="st">&quot;mypub.txt&quot;</span>)</span></code></pre></div>
<hr />
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/"><img alt="크리에이티브 커먼즈 라이선스" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png" /></a><br />이 저작물은 <a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/">크리에이티브 커먼즈 저작자표시-비영리-변경금지 4.0 국제 라이선스</a>에 따라 이용할 수 있습니다.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="basic-statistics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-science-and-tidyverse.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
